### PR TITLE
Elaborate on the semantics of `page_size`.

### PR DESCRIPTION
**Pull request by sunfishcode**
This fills out the specification of the `page_size` operator a little.

### CODE REVIEW

File: AstSemantics.md [hunk 1]
```diff
@@ -227,6 +227,11 @@ MVP, there are [future features](FutureFeatures.md#finer-grained-control-over-me
 proposed to allow setting protection and creating mappings within the
 contiguous linear memory.
 
+In the MVP, the result type of `page_size` is int32`; in the future, support
```
Comment (jfbastien) [inline AstSemantics.md]: One missing `backtick` here.

File: AstSemantics.md [hunk 2]
```diff
@@ -227,6 +227,11 @@ MVP, there are [future features](FutureFeatures.md#finer-grained-control-over-me
...
+support for page sizes requiring a type of `int64` may be added. The page size
+is an unsigned integer which is a power of 2 and at least 1024. `page_size`
```
Comment (jfbastien) [inline AstSemantics.md]: Mention that this is bytes?
Also, something along the lines of "it's not necessarily the actual page size, it's a suggested assumption from the runtime".

File: AstSemantics.md (hunk 1 above)
Comment (sunfishcode) [inline AstSemantics.md]: Fixed.

**Comment by titzer** [comment]
What can page_size be used for in the MVP?

**Comment by sunfishcode** [comment]
@titzer [The `resize_memory` operation](https://github.com/WebAssembly/design/blob/master/AstSemantics.md#resizing).

File: AstSemantics.md (hunk 2 above)
Comment (sunfishcode) [inline AstSemantics.md]: Fixed.

**Comment by titzer** [comment]
What if we just limited memory sizes to powers of 2?

**Comment by jfbastien** [comment]
@titzer going from 2GiB to 4GiB seems like a big jump :)

Though changing `page_size` to be `page_size_log2` would be interesting.

**Comment by titzer** [comment]
page_size = 1gb, problem solved :-)

**Comment by sunfishcode** [comment]
@titzer After thinking about it more, I agree. I removed the minimum page size.

**Comment by jfbastien** [comment]
lgtm

**Comment by titzer** [comment]
Sorry if I wasn't clear. What if we just limited memory sizes to powers of
2, which might obviate the need for page_size?

**Comment by sunfishcode** [comment]
On 32-bit devices, powers of 2 are very big steps relative to the underlying address space. A device may be able to easily allocate more than 128 MiB contiguous address space to a wasm app, but not all of 256 MiB, for example.

**Comment by jfbastien** [comment]
@titzer the OS we run on imposes a minimum page size. I guess we can over-allocate, and tell the user where memory now ends?

**Comment by lukewagner** [comment]
With asm.js, we started with a power of 2 allocation and got pushback that even the 128, 256, 512, 1024 quantas were too coarse, particularly on 32-bit, so we loosened the restriction to "multiple of 16mb" (for the benefit of ARM operand2 immediate encoding).  This power-of-2 limitation would be exacerbated in the many-small-wasm-modules-in-one-web-app use case we've discussed before.

**Comment by jfbastien** [comment]
Could we move this PR forward? I was looking for it in the repo, and just realized it was still uncommitted.

Small nit along the lines of #354, should this say "same type as the address' index"?

**Comment by jfbastien** [comment]
@sunfishcode reminded me that @titzer may want to chime in some more before we move this forward. Let's wait on him :)

**Comment by titzer** [comment]
I think we should just drop the language about int64 page sizes and
trapping, since I don't think there are any machines on the horizon that
don't fit the current requirements.

**Comment by sunfishcode** [comment]
WRT trapping, I agree. That's part of an older idea that is now removed.

WRT int64 page sizes, we can now reword this in terms of wasm32 and wasm64, which I think makes this clearer. What do you think of the updated text?

**Comment by titzer** [comment]
lgtm although the larger discussion around wasm32/wasm64 is probably still
not quite settled.

**Comment by jfbastien** [comment]
lgtms abound! Merging.

**Pull request merged**